### PR TITLE
Add per-viewer Component rendering

### PR DESCRIPTION
**Pull request by zml2008**
Supersedes #48 

This uses the `Pointer` system added in 4.8.0 to implement customizable rendering, based on things other than locale. The implementation is fairly similar to the original in #42, just removing `AudienceIdentity` entirely and replacing it with `Pointered`.

### CODE REVIEW

**Review by kezz — APPROVED** [review]
Do not let this incredibly minor set of nitpicks detract from the fact that this is an awesome PR. 🎉

File: platform-bukkit/src/main/java/net/kyori/adventure/platform/bukkit/BukkitAudiences.java [hunk 1]
```diff
@@ -91,5 +103,13 @@
    * @since 4.0.0
    */
   @NotNull Audience filter(final @NotNull Predicate<CommandSender> filter);
+
+  /**
+   * A builder for {@link BukkitAudiences}.
+   *
+   * @since 4.5.0
```
[the remark this answers is not in the document]
Comment (zml2008) [reply platform-bukkit/src/main/java/net/kyori/adventure/platform/bukkit/BukkitAudiences.java:110]: ah, forgot to clean that up from the original PR

File: platform-facet/src/main/java/net/kyori/adventure/platform/facet/FacetAudienceProvider.java [hunk 2]
```diff
@@ -23,18 +23,20 @@
...
 import net.kyori.adventure.audience.ForwardingAudience;
+import net.kyori.adventure.identity.Identity;
 import net.kyori.adventure.key.Key;
+import net.kyori.adventure.permission.PermissionChecker;
 import net.kyori.adventure.platform.AudienceProvider;
+import net.kyori.adventure.pointer.Pointered;
 import net.kyori.adventure.pointer.Pointers;
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
+import net.kyori.adventure.util.TriState;
```
[the remark this answers is not in the document]
Comment (zml2008) [reply platform-facet/src/main/java/net/kyori/adventure/platform/facet/FacetAudienceProvider.java:36]: sssh we don't talk about the imports

once this is merged, i'll hit optimize imports and then we'll pretend this never happened

File: platform-facet/src/main/java/net/kyori/adventure/platform/facet/Facet.java [hunk 3]
```diff
@@ -93,7 +97,8 @@
    * @return a facet or {@code null} if none are applicable
    * @since 4.0.0
    */
-  static <V, F extends Facet<V>> @Nullable F of(final @Nullable Collection<F> facets, final @Nullable V viewer) {
+  static <V, F extends Facet<V>> @Nullable F of(
```
Comment (kashike) [inline platform-facet/src/main/java/net/kyori/adventure/platform/facet/Facet.java:100]: bananas

File: platform-bukkit/src/main/java/net/kyori/adventure/platform/bukkit/BukkitAudience.java [hunk 4]
```diff
@@ -86,16 +88,35 @@
   );
   private static final Collection<Facet.Pointers<? extends CommandSender>> POINTERS = Facet.of(
     () -> new BukkitFacet.CommandSenderPointers(),
+    () -> new BukkitFacet.ConsoleCommandSenderPointers(),
     () -> new BukkitFacet.PlayerPointers()
   );
 
   private final @NotNull Plugin plugin;
+  // Bukkit only provides this as a String
+  private @Nullable Locale locale;
```
[the remark this answers is not in the document]
Comment (zml2008) [reply platform-bukkit/src/main/java/net/kyori/adventure/platform/bukkit/BukkitAudience.java:97]: Bukkit locale handling is entirely event-based -- we convert String -> Locale whenever we get an event, and cache it here in the audience. A sensible platform would cache the Locale for us, but this is Bukkit.

**Comment by valaphee** [comment]
This merge has mistakes, see https://github.com/KyoriPowered/adventure-platform/blob/master/platform-facet/src/main/java/net/kyori/adventure/platform/facet/FacetAudienceProvider.java#L114, getOrDefault doesn't exist in Audience, and plugins like PlotSquared doesn't work anymore:
```[16:54:03 ERROR]: Can not create Guice Injector8.9.2021 16:54:03 [16:54:03 ERROR]: java.lang.NoSuchMethodError: 'java.lang.Object net.kyori.adventure.platform.facet.FacetAudience.getOrDefault(net.kyori.adventure.pointer.Pointer, java.lang.Object)'8.9.2021 16:54:03 [16:54:03 ERROR]: java.lang.NoClassDefFoundError: Could not initialize class com.plotsquared.bukkit.util.BukkitUtil8.9.2021 16:54:03 [16:54:03 ERROR]: Error occurred while enabling PlotSquared v6.0.10-SNAPSHOT (Is it up to date?)```

**Comment by zml2008** [comment]
no, the only mistake is some plugin forgot to relocate

**Pull request merged**